remote-drivers/lwip-ethernet-async/server.c

```diff
@@ -0,0 +1,283 @@
+    /* mac address */
+    uint8_t hw_mac[6];
+    ps_io_ops_t *io_ops;
+
+    /* Pointers to shared_ringbuffers */
+    ring_handle_t *rx_ring;
+    ring_handle_t *tx_ring;
+} server_data_t;
+
+/* The camkes notification connection unregisters the callback each time. */
+static register_cb_fn reg_tx_cb;
+
+/**
+ * Packets have been transferred or dropped. Return the buffer
+ * into the available queue for reuse. 
+ * This is called by the actual driver.
+ *
+ * @param iface server data.
+ * @param cookie pointer to bufer descriptor to move.
+ *
+ */
+static void eth_tx_complete(void *iface, void *cookie)
+{   
+    ZF_LOGW("Packets have been transferred or dropped");
+    server_data_t *state = iface;
+
+    buff_desc_t *desc = cookie;
+
+    int err = enqueue_avail(state->tx_ring, desc->encoded_addr, desc->len, desc->cookie);
+    ZF_LOGF_IF(err, "lwip_eth_send: Error while enqueuing available buffer, tx available queue full");
+}
+
+/**
+ * Fetch an available buffer for receiving packets.   
+ * This is called by the actual driver.
+ *
+ * @param iface server data.
+ * @param buf_size size of buffer required. 
+ * @param cookie pointer to bufer descriptor to move.
+ *
+ * @return the physical memory address of the buffer. 
+ */
+static uintptr_t eth_allocate_rx_buf(void *iface, size_t buf_size, void **cookie)
+{
+    if (buf_size > BUF_SIZE) {
+        return 0;
+    }
+    server_data_t *state = iface;
+
+    uintptr_t addr;
+    size_t len;
+
+    /* Try to grab a buffer from the available ring */
+    if (driver_dequeue(state->rx_ring->avail_ring, &addr, &len, cookie)) {
+        ZF_LOGE("RX Available ring is empty. No more buffers available");
+        return 0;
+    }
+
+    void *decoded_buf = DECODE_DMA_ADDRESS(addr);
+    ZF_LOGF_IF(decoded_buf == NULL, "Decoded DMA buffer is NULL");
+
+    /* Invalidate the memory */
+    ps_dma_cache_invalidate(&state->io_ops->dma_manager, decoded_buf, buf_size);
+    uintptr_t phys = ps_dma_pin(&state->io_ops->dma_manager, decoded_buf, buf_size);
+
+    ZF_LOGW("phys: %p", phys);
+
+    return phys;
+}
+
+/**
+ * Packets have been received and are ready for processing. 
+ * Enqueue the packets into shared memory and notify the network stack. 
+ *
+ * @param iface server data.
+ * @param num_bufs the number of buffers filled with data.
+ * @param cookies an array of length num_bufs containing cookies passed in by eth_allocate_rx_buf()
+ * @param lens an array on length num_bufs containing the length of data each buffer now contains.  
+ *
+ */
+static void eth_rx_complete(void *iface, unsigned int num_bufs, void **cookies, unsigned int *lens)
+{
+    server_data_t *state = iface;
+    
+    for (int i = 0; i < num_bufs; i++) {
+        /* Add buffers to used rx ring. */
+        buff_desc_t *desc = cookies[i];
+        int err = enqueue_used(state->rx_ring, desc->encoded_addr, (size_t)lens[i], desc->cookie);
+        if (err) {
+            ZF_LOGE(err, "Enqueue failed: the RX used ring is full");
```

Comment:
This ZF_LOGE should probably be changed to ZF_LOGE_IFERR or the first parameter needs to be a string if the error isn't conditional
```suggestion
            ZF_LOGE("Enqueue failed: the RX used ring is full");
```